yasa/hypno.py

```diff
@@ -48,6 +24,826 @@
+        self._labels = labels
+        self._mapping = mapping
+        self._scorer = scorer
+
+    def __repr__(self):
+        return f"{self.hypno}"
+
+    def __str__(self):
+        return f"{self.hypno}"
+
+    @property
+    def hypno(self):
+        """The hypnogram values, stored in a :py:class:`pandas.Series`."""
+        # Q: Should this be called `hyp.stage`?
+        return self._hypno
+
+    @property
+    def n_epochs(self):
+        """The number of epochs in the hypnogram."""
+        return self._n_epochs
+
+    @property
+    def freq(self):
+        """The frequency resolution of the hypnogram. Default is '30s'"""
+        return self._freq
+
+    @property
+    def sampling_frequency(self):
+        """The sampling frequency (Hz) of the hypnogram."""
+        return self._sampling_frequency
+
+    @property
+    def start(self):
+        """The start date/time of the hypnogram. Default is None."""
+        return self._start
+
+    @property
+    def timedelta(self):
+        """A :py:class:`pandas.TimedeltaIndex` vector with the accumulated time difference of each
+        epoch compared to the first epoch."""
+        return self._timedelta
+
+    @property
+    def n_stages(self):
+        """The number of accepted stages in the hypnogram."""
+        return self._n_stages
+
+    @property
+    def labels(self):
+        """The accepted stage labels."""
+        return self._labels
+
+    @property
+    def mapping(self):
+        """A dictionary with the mapping from string to integer values."""
+        return self._mapping
+
+    @mapping.setter
+    def mapping(self, map_dict):
+        assert isinstance(map_dict, dict), "`mapping` must be a dictionary, e.g. {'WAKE': 0, ...}"
+        assert all([val in map_dict.keys() for val in self.hypno.unique()]), (
+            f"Some values in `hypno` ({self.hypno.unique()}) are not in `map_dict` "
+            f"({map_dict.keys()})"
+        )
+        if "ART" not in map_dict.keys():
+            map_dict["ART"] = -1
+        if "UNS" not in map_dict.keys():
+            map_dict["UNS"] = -2
+        self._mapping = map_dict
+
+    @property
+    def mapping_int(self):
+        """A dictionary with the mapping from integer to string values."""
+        return {v: k for k, v in self.mapping.items()}
+
+    @property
+    def scorer(self):
+        """The scorer name."""
+        return self._scorer
+
+    # CLASS METHODS BELOW
+
+    def as_int(self):
+        """Return hypnogram as integers.
+
+        The default mapping from string to integer is:
+
+        * 2 stages: {"WAKE": 0, "SLEEP": 1, "ART": -1, "UNS": -2}
+        * 3 stages: {"WAKE": 0, "NREM": 2, "REM": 4, "ART": -1, "UNS": -2}
+        * 4 stages: {"WAKE": 0, "LIGHT": 2, "DEEP": 3, "REM": 4, "ART": -1, "UNS": -2}
+        * 5 stages: {"WAKE": 0, "N1": 1, "N2": 2, "N3": 3, "REM": 4, "ART": -1, "UNS": -2}
+
+        Users can define a custom mapping:
+
+        >>> hyp.mapping = {"WAKE": 0, "NREM": 1, "REM": 2}
```

Comment:
It's supported via the `@mapping.setter` property. My thought was that:

1. For most users, asking them to instantiate the class with both the string stage values AND the mapping is going to be too much. Plus, I'd like to avoid having too many parameters in the Hypnogram class.
2. However for more advanced users there is still the (hidden) possibility to reset the mapping once and for all

Reply:
Yes agree 100% this makes sense.